### PR TITLE
Editorial: remove implicit wrapping/unwrapping of completion records

### LINKED ISSUE #253: Notate abstract ops that can and cannot throw
In streams, which is heavily promise-based, there are a lot of abstract operations that cannot throw. The ES spec has these too, in smaller number.

Of course, this kind of nothrow vs. throw characteristic can be contagious. If you only use nothrow abstract ops in your abstract op, then your abstract op is nothrow. Whereas if you use any throwing ops, you are now throwing. Unless you are certain for external reasons that the given throw operation will not throw in this context, in which case you might want to assert that it is a non-abrupt-completion.

There are a couple action items here, some of which can be done independently:
- Come up with a convention for asserting that the return value is never an abrupt completion, e.g. `Let _result_ be ! AbstractOp()`. Use it throughout. A lot of places in the spec currently use "perform" for this purpose, which is not defined anywhere.
- Introduce EMU support for notating with throws/nothrow. (Or perhaps, abrupt/noabrupt.) You can see this in action in Streams, e.g. https://streams.spec.whatwg.org/#rs-abstract-ops. For the source format, I'd suggest either `<emu-clause nothrow>` or `<emu-clause flair="nothrow">`.
- Go through the ES spec and update all abstract ops with whether or not they can throw. This could be done incrementally.
- Build automated checker tools to enforce the contagion rules, based on throw/nothrow and ! or ?.

### LINKED ISSUE #486: !? usage
#424 was closed though the remaining calls of OrdinaryGetOwnPropery were not updated to include !. So that probably needs to be fixed.

While closing @bterlson also mentioned that all abstract operations return a completion value and therefore must use ? or !.

That does not seem consistent with the draft, e.g., IsCallable is never prefixed.

There should probably be some more reliable set of rules.

### LINKED ISSUE #496: abstract operations don't always return Completion Records
5.2 "Algorithm Conventions" says:

> Calls to abstract operations return Completion Records.

This is not always true.

For instance, consider GetFunctionRealm: when it doesn't return an abrupt completion, it returns a Realm Record. This is not a Completion Record, nor can it be implicitly converted into one (see 6.2.2.2), because a Realm Record isn't an ECMAScript language value, and so can't be the [[Value]] of a Completion Record.

Other examples:
- CreateListFromArrayLike returns a List.
- NewDeclarativeEnvironment returns a Lexical Environment.
- PrepareForOrdinaryCall returns an Execution Context.

### LINKED ISSUE #497: eliminating implicit conversions to/from Completion Records
Recap:

Section 6.2.2.2 "Implicit Completion Values" says:

> Unless it is otherwise obvious from the context, an algorithm statement that returns a value that is not a Completion Record, such as: `Return "Infinity"` means the same thing as: `Return NormalCompletion("Infinity")`.

and

> Any reference to a Completion Record value that is in a context that does not explicitly require a complete Completion Record value is equivalent to an explicit reference to the [[Value]] field of the Completion Record value unless the Completion Record is an abrupt completion.

That is, there are implicit conversions to and from Completion Records.

Reaction:

In https://github.com/tc39/ecma262/issues/253#issuecomment-165063648, @rossberg-chromium said:

> The spec should not be optimised for maximum brevity but for maximum clarity. I think the implicit conversion from completion records to their values is actively harmful in that regard.

And in https://github.com/tc39/ecma262/issues/486#issuecomment-201677699, @bterlson said:

> There is a rule in 5.2 [sic] that allows us to treat a completion record as if it were a value but it's something I'd like to get away from.

In https://github.com/tc39/ecma262/issues/253#issuecomment-165134231, I agreed, and said:

> I've thought of an approach that I think would eliminate most of these conversions. I'll raise it in another issue when I have some time.

This post outlines that approach.  (I was hoping to submit a pull request, but now I'm not sure I'll have time in the near future.)

---

The basic idea is to do away with "normal" completion records.

That is, anywhere that an algorithm currently deals with a Completion Record whose [[Type]] is `normal`, instead just use whatever the [[Value]] of that record is. (You can ignore the [[Target]], it's always `empty`.)

Things that would disappear:
- All of 6.2.2.2.
- NormalCompletion()
- Completion()
- Steps of the form `If val is Completion Record, let val be val.[[Value]]` in the definition of ReturnIfAbrupt and "!".

Section 6.2.2, rather than defining "Completion Record", would define "AbruptCompletion Record".

For terminology, you could say something like: a "completion value" is either an "abrupt completion value" (i.e., an AbruptCompletion Record) or a "normal completion value" (i.e., `empty` or an ECMAScript language value).

The part of this approach that requires some work is to find all the algorithmic steps that currently deal with a Completion Record that might be a Normal Completion, and re-code them to deal with a 'completion value' that might be a 'normal completion value'. I'm guessing you could find all the places that might need to change by searching for `[[Type]]`, `[[Target]]`, and `[[Value]]`, but that's just a guess.

I can volunteer to do the work, but as I say above, I'm not sure how soon I could deliver, so someone else might like to take it on.

### LINKED ISSUE #1572: ! shorthand for completion record [[Value]] extraction is unnecessary
In [§5.2.3.4 ReturnIfAbrupt Shorthands](https://tc39.github.io/ecma262/#sec-returnifabrupt-shorthands), we define a shorthand `!` that's not really related to ReturnIfAbrupt. This shorthand asserts that the completion record is a normal completion and extracts its `[[Value]]`. But at the end of [§5.2.3.1 Implicit Completion Values](https://tc39.github.io/ecma262/#sec-implicit-completion-values), we say

> Any reference to a Completion Record value that is in a context that does not explicitly require a complete Completion Record value is equivalent to an explicit reference to the [[Value]] field of the Completion Record value unless the Completion Record is an abrupt completion.

I think we should either

1. remove this clause and use the `!` shorthand or explicit `[[Value]]` extraction throughout the spec, or
1. remove the `!` shorthand, its usages, and any explicit `[[Value]]` extraction

### LINKED ISSUE #1796: Sorting out completion records
I would like to get [completion records](https://github.com/tc39/ecma262/labels/completion%20records) sorted out. My current plan is to

1. Change the [[Value]] slot of Completion Records to allow it to hold any type of value, not just ECMAScript language value (thus resolving #496 via the first of the two solutions discussed in [this comment](https://github.com/tc39/ecma262/issues/496#issuecomment-249371481))
1. Remove the bit which asserts that all runtime semantics return a completion record
1. Label each abstract operation with the type it returns (in prose), so that non-completion-record-returning operations are not said to return completion records
1. Ensure all callers of an abstract operation consume the returned value properly: in particular, if the operation returns a completion record, it is explicitly treated as such, and if it does not, it is not
1. Remove the bit which allows implicit unwrapping of completion records

I'm going to prepare a PR which does this as soon as I have time.

### LINKED ISSUE #2645: bug in ClassDefinitionEvaluation
Step 31.b.ii is

```
ii. Let result be ? Call(elementRecord.[[BodyFunction]], F).
```

That `?` is stray - on the next line we check for abrupt completions so we can fix up the PrivateEnvironment. The `?` prevents us from getting there, so the PrivateEnvironment ends up wrong.

Fix is to just remove the `?`. We'll do that as part of #2547.

### PR DESCRIPTION
**Pull request by michaelficarra**
Ready for review. Based on #2546. Fixes #1796. Fixes #497. Fixes #496. Closes #1916. Fixes #253. Fixes #1572. Closes #486. Closes #2612. Fixes #2645.

1. [x] remove the provision for implicit wrapping in normal completion
1. [x] remove the provision for treating normal completions as their [[Value]]
1. [x] annotate return types
1. [x] make sure every AO has a return type specified; figure out what to do with any remaining explicit "unknown" returns
1. [x] see if we can replace any `~empty~` returns with `~unused~`
1. [x] make sure all internal methods with the same name have matching return types
1. [x] make sure all concrete methods with the same name have matching return types
1. [x] call sites for AOs that return completion records should use `?`/`!`/`Completion()`; call sites for AOs that do not return completion records should not use `!`/`?`/`Completion()`
1. [x] make sure that each AO that doesn't return a completion record also doesn't contain a step that uses `ReturnIfAbrupt`/`?`/`throw`
1. [x] make sure that each AO that does return a completion record might plausibly return an abrupt completion, i.e. it uses `?` or `throw` or `ReturnIfAbrupt` or returns a completion explicitly
1. [x] make sure that all AOs that return `~unused~` are always invoked using "Perform". if it's a completion record that normally contains `~unused~`, make sure only its `[[Type]]` is relied upon.
1. [x] If an AO is only ever invoked using "Perform", it should return `~unused~` or a completion record that normally contains `~unused~`
1. [x] find any AOs that return completion records but are always called using `!` and make them not return completion records
1. [x] add notational convention for wrapping of returns in `NormalCompletion` in AOs that return completion records; except `Completion` or `ThrowCompletion` or `NormalCompletion`
1. [x] replace `return Completion(X)` with  `return ? X` in AOs that return a completion record
1. [x] make sure that all uses of `Completion()` are on the call of an AO that returns a completion, field accesses, or some other opaque source
1. [x] go through each AO that returns a completion record with unknown normal type and see if we can further refine it
1. [x] make sure we're not returning `Completion(...)` or `Completion Record { ... }` or `NormalCompletion(...)` or `ThrowCompletion(...)` in algorithms that don't return completion records
1. [x] check as many of the above as possible in ecmarkup and bump ecmarkup
1. [x] go through effect suppressions and make sure they are still valid and necessary
1. [x] go through issues mentioned in #2548 and make sure they are addressed here
1. [ ] add visual indicator to AO headers when the AO returns a completion record
1. [x] explain what "a Completion Record normally containing an X" means; maybe think of better phrasing?

### CODE REVIEW

**Comment by arhadthedev** [comment]
The PR introduces a term <code>a Completion Record normally containing [a *Type*]</code>. I think it would be useful to explicitly define it in <https://tc39.es/ecma262/#sec-completion-record-specification-type> via a normal completion (probably with no `<dfn>` to avoid visual clutter). Otherwise it may seem that "normally" means "sometimes it can be violated and another type may appear".

File: spec.html [hunk 1]
```diff
@@ -845,29 +845,22 @@ <h1>Syntax-Directed Operations</h1>
...
-        <p>The abstract operation Completion(_completionRecord_) is used to emphasize that a previously computed Completion Record is being returned. The Completion abstract operation takes a single argument, _completionRecord_, and performs the following steps:</p>
+      <emu-clause id="sec-completion-ao" type="abstract operation">
+        <h1>
+          Completion (
+            _completionRecord_: a Completion Record,
+          ): a Completion Record
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>It is used to emphasize that a Completion Record is being returned.</dd>
```
Comment (arhadthedev) [inline spec.html:858]: In the light of explicit emphasis via a return type, does `Completion()` give any benefit or it may be removed altogether?

Comment (bakkot) [reply spec.html:858]: You don't have the return type sitting in front of you at the call site, only at the definition site, so yes, it's still helpful.

File: spec.html [hunk 2]
```diff
@@ -6285,15 +6285,16 @@ <h1>
           _O_: an Object,
           _P_: a property key,
           _V_: an ECMAScript language value,
-        ): a Completion Record normally containing a Boolean
+        ): a Completion Record normally containing ~unused~
```
Comment (ljharb) [inline spec.html:6288]: "normally containing" implies to me that sometimes it will contain something else. is this trying to convey that, or that it's a Completion of [[Type]] normal, containing `~unused~`?

Comment (bakkot) [reply spec.html:6288]: @ljharb the second thing; see [comment above](https://github.com/tc39/ecma262/pull/2547#issuecomment-940356606).

Comment (ljharb) [reply spec.html:6288]: thanks; the comment sums it up - altho i think i'd rather find another term than just settle for defining it.

File: spec.html [hunk 3]
```diff
@@ -867,7 +867,7 @@ <h1>
           <dt>description</dt>
           <dd>It is used to emphasize that a Completion Record is being returned.</dd>
         </dl>
-        <emu-alg>
+        <emu-alg example>
```
Comment (ljharb) [inline spec.html:870]: how is this an example?

Comment (michaelficarra) [reply spec.html:870]: oops, thanks for catching that

**Comment by jmdyck** [comment]
(You should probably also update the PR description where it says "Not ready for review".)

**Comment by jmdyck** [comment]
> replace `return Completion(X)` with `return ? X` in AOs that return a completion record

Why?

**Comment by michaelficarra** [comment]
@jmdyck No reason, we just thought it was stylistically neater and semantically it's equivalent.

**Comment by bakkot** [comment]
In particular, we already have lots of `Return ? Foo()`, which is equivalent to `Return Completion(Foo())`; we might as well pick one of those two forms and stick to it. And the former is less noisy (and, to my mind, clearer).

**Comment by jmdyck** [comment]
> In particular, we already have lots of `Return ? Foo()`, which is equivalent to `Return Completion(Foo())`; we might as well pick one of those two forms and stick to it.

Okay. Given those choices, I'd vote for `Return Completion(Foo())`.

> And the former is less noisy (and, to my mind, clearer).

I might agree that `Return ? X` is less noisy than `Return Completion(X)`, but
- not much, and
- if noise is a concern, you could invent a less noisy notation for `Completion(X)`.

As for "clearer", I strongly disagree. To me,
- `Return Completion(X)` says: `X` is a CR, which is returned.
- `Return ? X` says: `X` is a CR. If it's an abrupt CR, then it's [implicitly] returned (via the semantics of `?`).  Otherwise (i.e., if it's a normal CR), the CR's `[[Value]]` is extracted and becomes the argument of the `Return`, at which point that value is implicitly passed to `NormalCompletion()`, and the resulting normal CR is [explicitly] returned.

Also, in a lot of cases of `Return ? _cr_`, the `_cr_` is never a normal CR, so the abrupt 'path' is always taken and the `Return` is never actually executed, which feels odd. But there's no corresponding problem for `Return Completion(_cr_)`.

**Comment by bakkot** [comment]
It's true that if you think about what the spec looks like after applying the rewrite rules, that thing ends up being a fairly awkward sequence of steps.

But most readers are not thinking about what the spec looks like after applying rewrite rules, and they already encounter `? AO()` all the time. `Return ? AO()` looks like `Let _a_ be ? AO()` and requires no further thought. Using a different construct than the usual one here obscures that fundamental similarity.

I stand by "clearer" here.

**Comment by jmdyck** [comment]
> It's true that if you think about what the spec looks like after applying the rewrite rules, that thing ends up being a fairly awkward sequence of steps. But most readers are not thinking about what the spec looks like after applying rewrite rules,

I'm not thinking about what the spec looks like after applying rewrite rules, I'm thinking about the semantics of `?`.

> and they already encounter `? AO()` all the time.

I don't understand: are you saying they're used to seeing it, so they don't think about what the `?` means? Or they do, but in a simplified way?

> `Return ? AO()` looks like `Let _a_ be ? AO()` and requires no further thought. Using a different construct than the usual one here obscures that fundamental similarity.

By the same logic, `Return Completion(X)` looks like `Let _a_ be Completion(X)` and requires no further thought; using a different construct would obscure that fundamental similarity.

(Personally, I don't think the putative obscuring happens in either direction.)

File: spec.html [hunk 4]
```diff
@@ -18005,28 +18051,27 @@ <h2>Syntax</h2>
...
           1. Let _len_ be _nextIndex_ + 1.
           1. Perform ? Set(_array_, *"length"*, 𝔽(_len_), *true*).
-          1. NOTE: The above Set throws if _len_ exceeds 2<sup>32</sup>-1.
+          1. NOTE: The above step throws if _len_ exceeds 2<sup>32</sup>-1.
           1. Return _len_.
         </emu-alg>
         <emu-grammar>Elision : Elision `,`</emu-grammar>
         <emu-alg>
-          1. Return the result of performing ArrayAccumulation of |Elision| with arguments _array_ and _nextIndex_ + 1.
+          1. Return ? ArrayAccumulation of |Elision| with arguments _array_ and _nextIndex_ + 1.
```
Comment (jmdyck) [inline spec.html]:
```suggestion
          1. Return ? ArrayAccumulation of |Elision| with arguments _array_ and (_nextIndex_ + 1).
```
Otherwise, there's an ambiguity: an alternative parse would have `? ArrayAccumulation of |Elision| with arguments _array_ and _nextIndex_` as the left-hand operand of the `+`.

File: spec.html [hunk 5]
```diff
@@ -19158,7 +19191,7 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
           1. Let _firstSub_ be ? GetValue(_firstSubRef_).
           1. Let _restSub_ be ? SubstitutionEvaluation of |TemplateSpans|.
           1. Assert: _restSub_ is a possibly empty List.
-          1. Return the list-concatenation of &laquo; _firstSub_ &raquo; and _restSub_.
+          1. Return NormalComplation(the list-concatenation of &laquo; _firstSub_ &raquo; and _restSub_).
```
Comment (jmdyck) [inline spec.html]:
```suggestion
          1. Return NormalCompletion(the list-concatenation of &laquo; _firstSub_ &raquo; and _restSub_).
```

File: spec.html [hunk 6]
```diff
@@ -21095,29 +21129,31 @@ <h1>Runtime Semantics: Evaluation</h1>
...
+          1. Return ~empty~.
         </emu-alg>
         <emu-note>
           <p>A static semantics rule ensures that this form of |LexicalBinding| never occurs in a `const` declaration.</p>
         </emu-note>
         <emu-grammar>LexicalBinding : BindingIdentifier Initializer</emu-grammar>
         <emu-alg>
           1. Let _bindingId_ be StringValue of |BindingIdentifier|.
-          1. Let _lhs_ be ResolveBinding(_bindingId_).
+          1. <!-- TODO: Figure out if ResolveBinding can actually throw here. --> Let _lhs_ be Completion(ResolveBinding(_bindingId_)).
```
Comment (jmdyck) [inline spec.html]: Delete or resolve TODO comment?

File: spec.html [hunk 7]
```diff
@@ -29431,7 +29475,7 @@ <h1>
             1. For each element _nextKey_ of _keys_, do
               1. If Type(_nextKey_) is Symbol and _type_ is ~symbol~ or Type(_nextKey_) is String and _type_ is ~string~, then
                 1. Append _nextKey_ as the last element of _nameList_.
-            1. Return CreateArrayFromList(_nameList_).
+            1. Return _nameList.
```
Comment (jmdyck) [inline spec.html]:
```suggestion
            1. Return _nameList_.
```

File: spec.html [hunk 8]
```diff
@@ -41524,9 +41559,9 @@ <h1>
...
-          1. Perform LeaveCriticalSection(_WL_) and suspend _W_ for up to _timeout_ milliseconds, performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. _W_ can notify either because the timeout expired or because it was notified explicitly by another agent calling NotifyWaiter(_WL_, _W_), and not for any other reasons at all.
+          1. Perform LeaveCriticalSection(_WL_) and suspend _W_ for up to _timeout_ milliseconds, performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. _W_ can notify either because the timeout expired or because it was notified explicitly by another agent calling NotifyWaiter with arguments _WL_ and _W_, and not for any other reasons at all.
           1. Perform EnterCriticalSection(_WL_).
-          1. If _W_ was notified explicitly by another agent calling NotifyWaiter(_WL_, _W_), return *true*.
+          1. If _W_ was notified explicitly by another agent calling NotifyWaiter with arguments _WL_ and _W_, return *true*.
```
Comment (jmdyck) [inline spec.html]: Not sure why these were changed to SDO-style "with arguments". NotifyWaiter is invoked with a parenthesized list of args, so why wouldn't these references to hypothetical invocations be formatted that way too?

**Review by jmdyck — CHANGES_REQUESTED** [review]

**Comment by bakkot** [comment]
> I don't understand: are you saying they're used to seeing it, so they don't think about what the `?` means? Or they do, but in a simplified way?

It's not so much "simplified" as "abstracted": `X()` might be abrupt, in which case `Return ? X()` returns abruptly, and it might be normal, in which case it returns normally. If you think about this in terms of the concrete Record values which represent those two branches then you might find this strange, since it is after all going to have the same behavior in both branches, yes. But the whole point of `?` is that it abstracts away the concrete values in most cases, so that you don't actually have to think about them to follow the flow of the algorithm. 

> By the same logic, `Return Completion(X)` looks like `Let _a_ be Completion(X)` and requires no further thought; using a different construct would obscure that fundamental similarity.

Yes, and that's the point. `Let _a_ be Completion(X)` is an unusual thing to encounter and indicates something strange is going on: it means we're manipulating Completion values explicitly, usually because we're doing something with Promises or iterators which requires additional steps rather than just propagating abrupt completions. So you _should_ find `Let _a_ be Completion(X)` to be unusual; typical algorithm steps should _not_ be similar to it, even though we are in fact almost always manipulating Completion values. The point of abstractions like `?` is to background details like that.

By contrast, `Return ? X()` and `Let _a_ be ? X()` are the same kind of thing as each other: not working with a Completion directly, just propagating abrupt completions the same way we usually do.

File: package.json [hunk 9]
```diff
@@ -27,7 +27,7 @@
   "license": "SEE LICENSE IN https://tc39.es/ecma262/#sec-copyright-and-software-license",
   "homepage": "https://tc39.es/ecma262/",
   "dependencies": {
-    "ecmarkup": "^9.8.0"
+    "ecmarkup": "^9.8.1"
```
Comment (ljharb) [inline package.json:30]: since this is a patch version, the changes in this PR shouldn't be needed to land the ecmarkup upgrade, right?

Comment (bakkot) [reply package.json:30]: The patch fixes a bug which this PR would exercise, so yes, it is necessary.

File: spec.html [hunk 10]
```diff
@@ -1335,7 +1364,7 @@ <h1>Numeric Types</h1>
               BigInt::exponentiate
             </td>
             <td>
-              BigInt; may throw *RangeError*
+              a Completion Record normally containing a BigInt
```
Comment (ljharb) [inline spec.html]:
```suggestion
              a Completion Record (Abrupt, or Normal containing a BigInt)
```
what about this? the form also works for Return/Continue/Break

File: package.json (hunk 9 above)
Comment (michaelficarra) [reply package.json:30]: The question was whether we could do the ecmarkup upgrade first. This patch bump we could do, but the upcoming ecmarkup bump will be major, and should be included in this PR. So no point in doing this patch bump first.

**Comment by jmdyck** [comment]
When I try to "Load diff" for #2547, I get "Oops, something went wrong."
So instead, I've submitted my suggested changes (old ones and new ones) in [a PR against GH-1796 branch](https://github.com/tc39/ecma262/pull/2653).

**Comment by jmdyck** [comment]
Oh, also, I noticed that you have 4 occurrences like:
```a Record with [[A]] and [[B]] fields```
and 3 occurrences like:
```a Record with fields [[A]] (a Foo) and [[B]] (a Bar)```

The latter is more informative, so presumably preferable. Did you want to convert the former?

File: spec.html (hunk 8 above)
Comment (michaelficarra) [reply spec.html]: Two reasons:

1. We didn't want these references to be treated as calls by ecmarkup because (if they were) they're not used appropriately for the types.
2. There's actually ambiguity in the previous phrasing. Are we referring to the result of the call? This makes it clear that we are referring to `NotifyWaiter` itself.

File: spec.html (hunk 6 above)
Comment (michaelficarra) [reply spec.html:21142]: Deleted, but I think we should still try to figure this out.

File: spec.html (hunk 10 above)
Comment (bakkot) [reply spec.html:1367]: If everyone else prefers this form I'll defer to that. Let's get a vote from @syg?

Comment (michaelficarra) [reply spec.html:1367]: I think we should just wait until we put all options on the table at the editor call.

Comment (jmdyck) [reply spec.html:1367]: (It's a bit wordy, but on the other hand, you're clearly not going for conciseness, else you could say something like `Throw | Normal(BigInt)`.)

File: spec.html (hunk 8 above)
Comment (jmdyck) [reply spec.html]: Okay, make sense.

**Comment by jmdyck** [comment]
I've submitted a PR for a third batch of changes: #2657

Also, I noticed some other things but wasn't sure how to resolve them:

HasBinding:
Its return is "a Completion Record normally containing a Boolean", and yet almost all invocations of HasBinding() treat it like its return is just "a Boolean". (i.e., no `!`, `?`, or `Completion()`)

AsyncGeneratorUnwrapYieldResumption:
can't it return a normal completion?

There's an ambiguity:
`a List of Strings or *null*`
could mean:
`a List of (Strings or *null*)`
or:
`(a List of Strings) or *null*`

**Comment by bakkot** [comment]
> HasBinding

Oh, good catch. We went through all the concrete methods to make sure their declared types were consistent but we forgot to check the uses afterwards, I think.

I believe the sites which currently are missing `!` or `?` or `Completion` should all be `!`; that method only throws for object and global environment records, which I think are all screened out for the use sites in question. I'll go over it with @michaelficarra to check and update though.

> AsyncGeneratorUnwrapYieldResumption: can't it return a normal completion?

Oh, from the first step? Yup, I think you're right.

> There's an ambiguity:

Yeah, English has this problem. Probably best to add "either" to disambiguate, I guess.

**Comment by jmdyck** [comment]
(Diff still isn't working, so I'll just put this here.)
For `GetExportedNames`, in the return type,
> `either a List of Strings or *null*`

should be:
> `a List of either Strings or *null*`

**Comment by bakkot** [comment]
Ha, great catch, fixed. I think we read the ambiguity wrong while disambiguating, which is pretty good evidence for the need for `either`.

**Review by ljharb — COMMENTED** [review]
there's changes pending after this; i'll review those ne

File: spec.html [hunk 11]
```diff
@@ -7071,14 +7107,14 @@ <h1>
         CreateIterResultObject (
           _value_: unknown,
           _done_: a Boolean,
-        )
+        ): an Object that conforms to the <i>IteratorResult</i> interface
```
[the remark this answers is not in the document]
Comment (michaelficarra) [reply spec.html]: no https://github.com/tc39/ecma262/issues/2486

File: spec.html [hunk 12]
```diff
@@ -19008,9 +19051,7 @@ <h1>
           1. If Type(_func_) is not Object, throw a *TypeError* exception.
           1. If IsCallable(_func_) is *false*, throw a *TypeError* exception.
           1. If _tailPosition_ is *true*, perform PrepareForTailCall().
-          1. Let _result_ be Call(_func_, _thisValue_, _argList_).
-          1. Assert: If _result_ is not an abrupt completion, then Type(_result_) is an ECMAScript language type.
```
[the remark this answers is not in the document]
Comment (jmdyck) [reply spec.html]: > if completions can only contain language values

This PR has lots of examples of normal completions containing non-language values (e.g., a List, a Record). However, in the "Completion Record Fields" table, `[[Value]]` still says "an ECMAScript language value or ~empty~", so that should probably be fixed.

[the remark this answers is not in the document]
Comment (michaelficarra) [reply spec.html]: Completion Records can contain things that are not ECMAScript language values. `~unused~` is a very common one. This assertion is removed because it's redundant with the type signature.

[the remark this answers is not in the document]
Comment (ljharb) [reply spec.html]: ah, because `Call` can only return a normal completion when it contains an ES language value, got it. thanks!

File: spec.html [hunk 13]
```diff
@@ -855,72 +855,65 @@ <h1>Syntax-Directed Operations</h1>
...
+          Completion (
+            _completionRecord_: a Completion Record,
+          ): a Completion Record
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>It is used to emphasize that a Completion Record is being returned.</dd>
+        </dl>
         <emu-alg>
           1. Assert: _completionRecord_ is a Completion Record.
```
[the remark this answers is not in the document]
Comment (ljharb) [reply spec.html:871]: i guess so, but i thought the main benefit of the structured param types was to remove the explicit assertion

[the remark this answers is not in the document]
Comment (bakkot) [reply spec.html:871]: In general, yes, because in general the assertion is incidental to the operation of the AO. Here the assertion is the whole point of the AO.

[the remark this answers is not in the document]
Comment (ljharb) [reply spec.html:871]: Fair enough!

**Comment by syg** [comment]
My review since I can't get GitHub to load the diff at all.

Without listing every location, the "either a normal completion containing either an X or a Y or an abrupt completion" doesn't read great. IMO this is because of the nested "either", so I recommend for those to be changed to:

2-form: either a normal completion containing an X or Y, or an abrupt completion

2+-form: either a normal completion containing an X, Y, or Z, or an abrupt completion

<hr>

L964: `<p>The Evaluation syntax-directed operation and all built-in functions return a Completion Record. It is an editorial error if a Completion Record is returned from such an abstract operation through any other means. For example, within these abstract operations,</p>`

I find the "The Evaluation [...]" sentence out of place and also makes "such an abstract operation" confusing in the following sentence. While it is true that Evaluation and built-ins always return completions, this section is about the _implicit_ normal completions. Let's move it up to L958 as:

In algorithms within abstract operations which are declared to return a Completion Record, **within the Evaluation syntax-directed operation, and within all built-in functions**, the returned value is first passed to NormalCompletion, and the result is used instead. This rule does not apply within the Completion algorithm or when the value being returned is clearly marked as a Completion Record in that **same** step; these cases are:

<hr>

L985: Please add a short explanation for why the example is an editorial error.

<hr>

L9493: Missing `?`, should be `1. Return ? _result_.`
L21688: Missing `?`, should be `1. Return ? _bodyResult_.`

<hr>

L38842: Nice, suppressing UC here.

<hr>

L41573: Why was the `with arguments _WL_ and _W_` change made?
L41575: Ditto

**Comment by michaelficarra** [comment]
@syg

> L41573: Why was the `with arguments _WL_ and _W_` change made?

See https://github.com/tc39/ecma262/pull/2547#discussion_r799945817

**Comment by syg** [comment]
> See https://github.com/tc39/ecma262/pull/2547#discussion_r799945817

Ah okay, makes sense.

**Comment by bakkot** [comment]
@ljharb @syg I believe we've addressed all your comments. The first commit is just the rebase and the subsequent ones are changes we made after your reviews; PTAL.

File: spec.html (hunk 8 above)
Comment (michaelficarra) [reply spec.html]: @ljharb This is another little thing that can be pulled into a separate PR if you like.

Comment (bakkot) [reply spec.html]: It would need to land first, though.

Comment (ljharb) [reply spec.html]: #2667

**Review by syg — APPROVED** [review]
Wording lgtm, trusting correctness to existing checks.

File: spec.html [hunk 14]
```diff
@@ -24765,15 +24809,15 @@ <h1>
...
           1. Let _F_ be _constructorInfo_.[[Closure]].
-          1. Perform ! MakeClassConstructor(_F_).
-          1. Perform ! SetFunctionName(_F_, _className_).
-        1. Perform ! MakeConstructor(_F_, *false*, _proto_).
+          1. Perform MakeClassConstructor(_F_).
+          1. Perform SetFunctionName(_F_, _className_).
+        1. Perform MakeConstructor(_F_, *false*, _proto_).
         1. If |ClassHeritage?| is present, set _F_.[[ConstructorKind]] to ~derived~.
-        1. Perform ! CreateMethodProperty(_proto_, *"constructor"*, _F_).
+        1. Perform CreateMethodPropertyInfallibly(_proto_, *"constructor"*, _F_).
```
[the remark this answers is not in the document]
Comment (michaelficarra) [reply spec.html]: Already done.

File: spec.html [hunk 15]
```diff
@@ -29461,23 +29512,23 @@ <h1>Object.getOwnPropertyDescriptors ( _O_ )</h1>
         <h1>Object.getOwnPropertyNames ( _O_ )</h1>
         <p>When the `getOwnPropertyNames` function is called, the following steps are taken:</p>
         <emu-alg>
-          1. Return ? GetOwnPropertyKeys(_O_, ~string~).
+          1. Return CreateArrayFromList(? GetOwnPropertyKeys(_O_, ~string~)).
```
[the remark this answers is not in the document]
Comment (michaelficarra) [reply spec.html]: Yeah just a small refactor. See my comment over there.

**Comment by michaelficarra** [comment]
@ljharb Addressed. Thanks for the review.
